extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java

```diff
@@ -162,219 +166,216 @@ public synchronized void setError(File[] error) {
-                                        if (!STACK_TRACE.matcher(str).find()) {
-                                            StandardLogger.findHyperlink(str, session, null).println(session, err);
-                                        }
-                                    } else {
-                                        // do not match long strings, directly create a trivial hyperlink
+                                    if (str.length() >= LOGGER_MAX_LINE_LENGTH || !STACK_TRACE.matcher(str).matches())
                                         StandardLogger.findHyperlink(str, session, null).println(session, err);
-                                    }
                                     log(str, logLevel);
-                                    currentLine.reset();
-                                } else {                                    
-                                    currentLine.write(c);
-                                    if(currentLine.size() > 8192) {
-                                        flusher.run();
-                                    } else {
-                                        flusher.schedule(250);
-                                    }
                                 }
-                            }    
+                                last = i + 1;
+                            }
                         }
+
+                        if (last < len)
+                            append(data, last, len - last);
                     }
                 } finally {
-                    if (logLevel != null) {
-                        maybeFlush();
-                        if (err) {
-                            foldingHelper.clearHandle();
-                        }
-                    }
+                    maybeFlush(true);
                 }
             } catch (IOException x) {
                 // ignore IOException: Broken pipe from FileOutputStream.writeBytes in BufferedOutputStream.flush
-            } catch (ThreadDeath d) {
-                // OK, build just stopped.
-                return;
             }
-            //System.err.println("copied " + in + " to " + out + "; content='" + content + "'");
         }
 
-        private synchronized void maybeFlush() {
-            if (ow == null) { // ?? #200365
-                return;
-            }
+        public void maybeFlush(boolean end) {
             try {
-                if (currentLine.size() > 0) {
-                    String str = currentLine.toString(encoding);
-                    ow.write(str);
-                    log(str, logLevel);
+                String str = take();
+                synchronized (foldingHelper) {
+                    if (!str.isEmpty()) {
+                        ow.write(str);
+                        log(str, logLevel);
+                    }
+                    if (end && err)
+                        foldingHelper.clearHandle();
                 }
```

Comment:
i like that you use `foldingHelper` as monitor instead of the task. This might fix a race condition, although interleaving `out` and `err` wouldn't result in proper folding anyway most of the time I suppose.

Reply:
It's a bit ugly but while i was cleaning up the final patch I noticed the comment on FoldingHelper stated it was supposed to be used as a lock for interleaving output, but then it wasn't actually used.  

With this changed it will ensure stderr/stdout has full line outputs before interleaving unless the timeout/flush happens in-between.  The output from the case below is improved although each line still interleaves due to the output rate.

    public class SlowOutput {
    static void deepStackTrace(int i) {
        if (i == 0) {
            try {
                System.out.println("/home/notzed/.emacs:5:4: yo ho");
                throw new Exception();
            } catch (Exception x) {
                x.printStackTrace(System.err);
            }
        } else {
            deepStackTrace(i - 1);
        }
    }
    public static void main(String[] args) throws InterruptedException {
        for (int i = 0; i < 10000; i++) {
            for (int j = 0; j < 1000; j++) {
                System.out.print(j);
            }
            System.out.println();
            if (i % 50 == 49)   deepStackTrace(5);
        }
    }
    }

Reply:
good stuff!